ds3-autogen-c/src/main/java/com/spectralogic/ds3autogen/c/CCodeGenerator.java

```diff
@@ -59,121 +62,119 @@ public CCodeGenerator() {
+    public static ImmutableList<Enum> getAllEnums(final Ds3ApiSpec spec) throws ParseException {
+        final ImmutableList.Builder<Enum> allEnumsBuilder = ImmutableList.builder();
+        if (ConverterUtil.hasContent(spec.getTypes())) {
             for (final Ds3Type ds3TypeEntry : spec.getTypes().values()) {
-                if (ConverterUtil.hasContent(ds3TypeEntry.getElements())) {
-                    final Type typeEntry = TypeConverter.toType(ds3TypeEntry);
-                    generateTypeTemplate(typeEntry, "TypeElement.ftl");
-                    generateTypeTemplate(typeEntry, "FreeTypeElementPrototype.ftl");
-                    generateTypeTemplate(typeEntry, "FreeTypeElement.ftl");
-                }
+                allEnumsBuilder.add(EnumConverter.toEnum(ds3TypeEntry));
             }
         }
+        return allEnumsBuilder.build();
+    }
 
-        // Generate Requests last because they depend on the code generated above
-        if (ConverterUtil.hasContent(spec.getRequests())) {
-            for (final Ds3Request request : spec.getRequests()) {
-                generateRequest(request);
+    public static ImmutableList<Struct> getAllStructs(final Ds3ApiSpec spec) throws ParseException {
+        final ImmutableList.Builder<Struct> allStructsBuilder = ImmutableList.builder();
+        if (ConverterUtil.hasContent(spec.getTypes())) {
+            for (final Ds3Type ds3TypeEntry : spec.getTypes().values()) {
+                allStructsBuilder.add(StructConverter.toStruct(ds3TypeEntry));
             }
         }
+        return allStructsBuilder.build();
     }
 
-    public void generateRequest(final Ds3Request ds3Request) throws IOException, TemplateException {
-        Template requestTemplate = null;
-        Request request = null;
-
-        if (ds3Request.getClassification() == Classification.amazons3) {
-            request = RequestConverter.toRequest(ds3Request);
-            requestTemplate = config.getTemplate("AmazonS3InitRequestHandler.ftl");
-        } else if (ds3Request.getClassification() == Classification.spectrads3) {
-            LOG.info("AmazonS3 request");
-            // TODO
-        } else if (ds3Request.getClassification() == Classification.spectrainternal) /* TODO && codeGenType != production */ {
-            LOG.debug("Skipping Spectra internal request");
-            return;
-        } else {
-            throw new TemplateException("Unknown dDs3Request Classification: " + ds3Request.getClassification().toString(), Environment.getCurrentEnvironment());
-        }
+    // return structs which contain only primitive types first, and then cascade for structs that contain other structs
+    public static ImmutableList<Struct> getStructsOrderedList(final Ds3ApiSpec spec) throws ParseException {
+        final ImmutableList.Builder<Struct> orderedStructsBuilder = ImmutableList.builder();
+        final Queue<Struct> allStructs = new LinkedList(getAllStructs(spec).asList());
```

Comment:
do you have to call `asList()` here?  As far as I know, ImmutableList implements the List interface.